### ISSUE TITLE
Numerical stability of attention layers

### ISSUE THREAD

**Issue by KhrulkovV** [lucidrains/imagen-pytorch#69]
Hello!

Thank's for a cool implementation!
I've been playing around with the code on CIFAR-10 trying to replicate SOTA fid values. However, I get some nasty loss explosions all of a sudden in the middle of training in fp16 with sensible configs, e.g., 
```
class BaseUnet32(Unet):
    def __init__(self, *args, **kwargs):
        kwargs.update(
            dict(
                dim=192,
                dim_mults=(1, 2, 4),
                num_resnet_blocks=2,
                layer_attns=(False, True, True),
                layer_cross_attns=False,
                attn_heads=1,
                ff_mult=2.0,
            )
        )
        super().__init__(*args, **kwargs)
```
I've looked into attention implementation in `guided-diffusion` and it seems that they do scaling by 1/sqrt(dim) a bit differently by redistributing it as 1/sqrt(sqrt(dim)) in q and k [(link)](https://github.com/openai/guided-diffusion/blob/27c20a8fab9cb472df5d6bdd6c8d11c8f430b924/guided_diffusion/unet.py#L386), as well as doing `softmax` on float32 inputs.
Testing it at the moment, will let you know if it helps

**Comment by lucidrains** [comment]
@KhrulkovV :wave: sure, let us try it! https://github.com/lucidrains/imagen-pytorch/commit/7c5e40ef1e6872c06d892a305edfca6f3e4f0d66 (i don't think distributing the scale between the q and k is necessary, as i already pre-scale the q's)

**Comment by lucidrains** [comment]
@KhrulkovV there's a few other tricks i can reach for if you are still running into issues (pb-relax from cogview, cosine sim attention, etc)

**Comment by lucidrains** [comment]
keep us posted on your experimental results!

**Comment by KhrulkovV** [comment]
Sadly, it still diverged ;(
<img width="405" alt="Screen Shot 2022-06-21 at 11 37 53" src="https://user-images.githubusercontent.com/7949649/174755947-c4399866-bf7e-4e27-9a73-cfd84e237ac6.png">

**Comment by KhrulkovV** [comment]
It did not fix it ;(

**Comment by lucidrains** [comment]
@KhrulkovV how high is your learning rate?

**Comment by lucidrains** [comment]
@KhrulkovV ohh i see what is going on, you need to do at least 4 attention heads (you are only using 1)

the reason why we do multi-headed attention is for the stability issues you are seeing

**Comment by KhrulkovV** [comment]
Hmm I ll try it, thanks

**Comment by lucidrains** [comment]
@KhrulkovV yeah, learning rate of `1e-3` is too high, you'll want to lower that to `1e-4`

**Comment by lucidrains** [comment]
@KhrulkovV perhaps that's why your other attention run diverged here https://github.com/lucidrains/memory-efficient-attention-pytorch/issues/2#issuecomment-1157529760

**Comment by KhrulkovV** [comment]
@lucidrains Hm I think I had learning rate 3e-4 and 8 heads in that run 😔 After that I started replicating cifar10 setup from Karras’s recent paper (they use 1e-3)

**Comment by KhrulkovV** [comment]
@lucidrains Hm I think I had learning rate 3e-4 and 8 heads in that run 😔 After that I started replicating cifar10 setup from Karras’s recent paper (they use 1e-3)

**Comment by lucidrains** [comment]
@KhrulkovV was that other flash-attention run done in float16 or float32?

**Comment by KhrulkovV** [comment]
> @KhrulkovV was that other flash-attention run done in float16 or float32?

They have only fp16 implementation for attention, so it was all in fp16

**Comment by lucidrains** [comment]
@KhrulkovV ohh got it, thanks!

**Comment by KhrulkovV** [comment]
But the boost was really nice - 1200 img/sec to 1800 img/sec until it diverged 😹

**Comment by lucidrains** [comment]
@KhrulkovV :laughing: yeahh, i've been chatting with a few other researchers, and the constant running renormalization may be causing issues. but i still think with some more research time, there could be a solution :thinking:

**Comment by lucidrains** [comment]
@KhrulkovV how did it go?

**Comment by KhrulkovV** [comment]
@lucidrains Hey, I set attention heads to 4 and lowered learning rate, going good so far

**Comment by lucidrains** [comment]
@KhrulkovV noice :^)
